### ISSUE TITLE
Modal: Refresh (via dev menu) does not tear down

### ISSUE THREAD

**Issue by pstanton** [facebook/react-native#17986]
### Is this a bug report?

yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

kind of

### Environment

Environment:
OS: Windows 10
Node: 9.0.0
Yarn: 1.3.2
npm: 5.5.1
Watchman: Not Found
Xcode: N/A
Android Studio: Version 3.0.0.0 AI-171.4408382

Packages: (wanted => installed)
react: 16.2.0 => 16.2.0
react-native: 0.53.0 => 0.53.0

  Target Platform: Android 7.0 on Samsung Galaxy S5

### Steps to Reproduce

1. create app and launch with `react-native run-android`
2. trigger state which causes a modal to show
3. launch the dev menu (via shake or keycode)
4. select refresh

### Expected Behavior

The app closes the modal and refreshes

### Actual Behavior

The app presumably refreshes, but the modal stays on top and cannot be closed

### Reproducible Demo

```

import React, { Component } from 'react';
import { Text, View, Button, Modal, StyleSheet } from 'react-native';

export default class MyComponent extends Component {
  state = {
    modalVisible: false,
  };

  openModal() {
    this.setState({modalVisible:true});
  }

  closeModal() {
    this.setState({modalVisible:false});
  }

  render() {
    return (
        <View style={styles.container}>
          <Modal
              visible={this.state.modalVisible}
              animationType={'slide'}
              onRequestClose={() => this.closeModal()}
          >
            <View style={styles.modalContainer}>
              <View style={styles.innerContainer}>
                <Text>This is content inside of modal component</Text>
                <Button
                    onPress={() => this.closeModal()}
                    title="Close modal"
                >
                </Button>
              </View>
            </View>
          </Modal>
          <Button
              onPress={() => this.openModal()}
              title="Open modal"
          />
        </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  modalContainer: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: 'grey',
  },
  innerContainer: {
    alignItems: 'center',
  },
});
```

**Comment by allpwrfulroot** [comment]
Seeing in iOS as well

**Comment by scottmas** [comment]
I can confirm this is happening on both Android and iOS

**Comment by MrLoh** [comment]
Is there any workaround? This also affects programmatic restarting on codepush updates for example.

**Comment by hanselsen** [comment]
I too wish there was a solution for this

**Comment by jjercx** [comment]
My workaround: 

Instead of this:
`<Modal visible{this.state.isVisible} />`

Do this:
`{this.state.isVisible && <Modal visible{true} />`

**Comment by ilovett** [comment]
This is pretty awful for developing against, but it is still possible to use the modal, so long as you close it before you reload the app.

This means you must have a way of closing the modal from the modal itself (IE: a close button), and you must close it before the app reloads...  you should also disable live reload and hot reload.

**Comment by micalm** [comment]
Can confirm this still happens on 0.55.4, both on a physical Xiaomi Redmi Note 4 (Android 7.0/MIUI Global 9.5) and an emulated API 23 device.

**Comment by jamesone** [comment]
Can confirm this is happening on Android and iOS using react-native 0.55.4

**Comment by theapache64** [comment]
`"react-native": "0.55.4"` contains this bug. CONFIRMED!

**Comment by lefrankleal** [comment]
@jjercx not working for me :/

**Comment by victorvhpg** [comment]
"react-native": "0.55.4"
   this is happening  here :(
![2018_06_23_17_11_53](https://user-images.githubusercontent.com/333451/41813263-e956c176-7708-11e8-8b95-d47c81aada71.gif)

**Comment by Montaguti** [comment]
Still presenting this behavior on `"react-native": "0.55.4"`

**Comment by MarcelZOI** [comment]
same here. 
react-native: "0.55.4"
any news?

**Comment by Thyoity** [comment]
Yes, this is annoying... Same thing happening here...

**Comment by PiotrKujawa** [comment]
I guess here's the fix https://github.com/facebook/react-native/commit/e5c2a66 and will be available in 0.56

**Comment by gwidener** [comment]
That fix is also in 0.55.4, and the problem remains (Android emulation at least).  It's a blocker for hot/live reloading, therefore for any decent development environment.  Can't use this package until it's fixed.

**Comment by chadhobson** [comment]
So.... what's the workaround here? This issue is 6 months old and its still happening on 56.0.

**Comment by george-steel** [comment]
I can also confirm this still happens on 0.56.

**Comment by sunnylqm** [comment]
ping @mdvacca @kelset

**Comment by itbeauregard** [comment]
Our project just updated from RN 0.50 to 0.56. I can confirm that this has become a frustrating issue for our app.

**Comment by ayoubgdah** [comment]
Still occurring in 0.56

**Comment by mjmasn** [comment]
Still occurring in 0.57.0

```
  React Native Environment Info:
    System:
      OS: Linux 4.15 Ubuntu 16.04.5 LTS (Xenial Xerus)
      CPU: x64 Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
      Memory: 204.00 MB / 15.54 GB
      Shell: 4.3.48 - /bin/bash
    Binaries:
      Node: 8.11.0 - ~/.nvm/versions/node/v8.11.0/bin/node
      Yarn: 1.9.4 - /usr/bin/yarn
      npm: 5.6.0 - ~/.nvm/versions/node/v8.11.0/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      Android SDK:
        Build Tools: 23.0.1, 26.0.1, 26.0.2, 26.0.3, 27.0.0, 27.0.3
        API Levels: 16, 19, 22, 23, 26, 27, 28
    IDEs:
      Android Studio: 3.1 AI-173.4907809
    npmPackages:
      react: ^16.5.0 => 16.5.1 
      react-native: 0.57.0 => 0.57.0 
    npmGlobalPackages:
      create-react-native-app: 1.0.0
      react-native-cli: 2.0.1
      react-native-git-upgrade: 0.2.7
```

**Comment by devpro9219** [comment]
I have confirmed .54, .55 , .56. 
Still has this issues.

**Comment by gaguirre** [comment]
I think this issue is more likely related to the **Modal** itself than the **Dev Menu**, so the solution should be addressed in the former.

However, what was really pissing me off right now was the **Dev Menu** so I've managed to workaround that specific issue touching some lines.
**It's not a solution by any means** and it works in iOS only, but may be helpful for someone else.

In the `RCTDevMenu.m` file, change the `toggle` method to
```
- (void)toggle
{
  UIViewController* presentedViewController = RCTPresentedViewController();
  if (_actionSheet) {
    [_actionSheet dismissViewControllerAnimated:YES completion:^(void){}];
    _actionSheet = nil;
  } else if ([presentedViewController class] == NSClassFromString(@"RCTModalHostViewController")) {
    [a dismissViewControllerAnimated:false completion:^{
      [self show];
    }];
  } else {
    [self show];
  }
}
```
This way, if a modal is present, it will be closed before opening the **Dev Menu**.

**Comment by barghi** [comment]
I have this issue in react native 0.57.2

**Comment by khuramdogar** [comment]
I also have this issue in 0.57.2. What should we use as alternative

**Comment by george-steel** [comment]
@khuramdogar I'm using [react-native-navigation](https://github.com/wix/react-native-navigation/tree/v2), which implements modals as well as a few other forms of navigation. It does not have this problem since it has its own native implementation. I started using it due to this specific bug.

**Comment by SiSa68** [comment]
Still have this issue in RN 0.57.4
Why?!!

**Comment by khuramdogar** [comment]
No Issue is resolved thanks for asking

**Comment by solkaz** [comment]
I am still seeing this issue as of 0.57.1

**Comment by ericwooley** [comment]
I wrote this and it solves my issue on android, haven't tried on ios yet.
```ts
import React from 'react'
import { Modal } from 'react-native'
type ExtractProps<
  TComponentOrTProps
> = TComponentOrTProps extends React.Component<infer TProps, any>
  ? TProps
  : TComponentOrTProps

export class UnMountingModal extends React.PureComponent<
  ExtractProps<Modal>,
  { mounted: boolean }
> {
  public state = { mounted: false }
  public componentDidMount() {
    this.setState({ mounted: true })
  }
  public componentWillUnmount() {
    this.setState({ mounted: false })
  }
  public render() {
    return (
      <Modal {...this.props} visible={this.state.mounted && this.props.visible}>
        {this.props.children}
      </Modal>
    )
  }
}
```

**Comment by tmswartz12** [comment]
Still having this issue! has anyone seen a solution?

**Comment by choo737** [comment]
Still having this issue here, RN0.55.4

**Comment by sunnylqm** [comment]
@PiotrKujawa Actually the [pr you mentioned](https://github.com/facebook/react-native/commit/e5c2a66897b9c562c549e63adcf70783ea34c418) is not the fix, but the cause! 

The root cause is [the removal of this line](https://github.com/facebook/react-native/commit/e5c2a66897b9c562c549e63adcf70783ea34c418#diff-6453d554bf967fe816a1aa3735c9fe62L172)
When we do reload, actually we [do a pause](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java#L1090), so the `dismiss()` in `onHostPause()` can be triggered. But since this pr merged since 0.53, so that is why this bug emerged since then.

@mdvacca Could you recall why you removed this line? I could not see how that was related to `Modal not disappearing when navigating to another activity`?

**Comment by sunnylqm** [comment]
Ideally, modal should be notified when ReactContext was tear down. The current flow only deal with main rootviews and pure native modules, but modal was a non typical rootview, neither recorded in ReactInstanceManager nor in NativeModuleRegistry, so it is forgotten in the between. 
Actually some other native modules also forgott to handle instance destroyed event, like Alert(DialogModule). If you enable live reload and fire an alert action, then modify the source code, you can see the Alert window also hang on the screen after reload.
@shergin Curious that has the new architecture considered these situations? IMHO,
1. the react instance events should be notifyed to viewmanagers as well, or,
2. there should be another list to manager special views outside of main rootviews in instance manager

**Comment by sunnylqm** [comment]
For now, I suggest using a js modal like react-native-root-modal instead of this one as a workaround.

**Comment by hurleyparanormal** [comment]
What is this 3

**Comment by noahtallen** [comment]
Confirming this is still an issue on the latest release, 0.57.8. Though the issue is very apparent when hot-reloading, it's also still an issue with any sort of reload while the modal is open.

**Comment by LRNZ09** [comment]
I'm also confirming this is still an issue on `0.57.8`

**Comment by Brad19** [comment]
I'm confirming it as well on 0.57.8. Could someone pls advise as our App is likely to get this issue soon and might impact customers once one of the features for the modal goes live.

**Comment by george-steel** [comment]
@rogueturnip I'm using react-native-navigation for this is it replaces the native initialization and teardown routines with ones that handle modals correctly.

**Comment by rogueturnip** [comment]
@george-steel Thanks for this!

**Comment by NahaseanNjenga** [comment]
Does anyone have a concrete solution for this? Am using RN 0.58.3

**Comment by rafaelfndev** [comment]
Still an issue on react-native `0.59.3`

**Comment by cdesch** [comment]
(The issue is present in Expo SDK `v32.0.0` which is based on RN `0.57`)

**Comment by nishantWM** [comment]
still facing this issue. Any updates ?

**Comment by saddacracker** [comment]
On emulator refresh, it appears it rendering a duplicate modal. I have to close the app in the emulator and relaunch to work around this.

**Comment by HVCode** [comment]
> On emulator refresh, it appears it rendering a duplicate modal. I have to close the app in the emulator and relaunch to work around this.

same here, "react-native": "0.59.8",

**Comment by Normence** [comment]
Still exists now!

**Comment by xXFracXx** [comment]
Still an issues.

**Comment by nishantWM** [comment]
> @PiotrKujawa Actually the [pr you mentioned](https://github.com/facebook/react-native/commit/e5c2a66897b9c562c549e63adcf70783ea34c418) is not the fix, but the cause!
> 
> The root cause is [the removal of this line](https://github.com/facebook/react-native/commit/e5c2a66897b9c562c549e63adcf70783ea34c418#diff-6453d554bf967fe816a1aa3735c9fe62L171)
> 
> ```diff
>   public void onHostPause() {	  
> -    // We dismiss the dialog and reconstitute it onHostResume
> -    dismiss();	
>   }
> ```
> 
> When we do reload, actually we [do a pause](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java#L1090), so the `dismiss()` in `onHostPause()` can be triggered. But since this pr merged into 0.53, there was no `dismiss` anymore. This bug emerged since then.
> 
> After adding that line back, now the modal can tear down normally
> ![modal](https://user-images.githubusercontent.com/615282/49095877-5df9a400-f2a4-11e8-979d-8de2833dd99f.gif)
> 
> @mdvacca Could you recall why you removed this line? I could not see how that was related to `Modal not disappearing when navigating to another activity`? Do you expect the modal should persist when navigation to another activity?

Removing this line helps. Any ideas why this is required?

**Comment by rajatgng** [comment]
please then somwone tell how to fix it

**Comment by AmruthPillai** [comment]
Is there no way to fix this? Still an issue :(

**Comment by Inovassist-dev** [comment]
Are you guys kidding that, after so many RN versions and month since this issue was open, nobody made a fix?

**Comment by EhsanSarshar** [comment]
you can set the visibility of modal to false in component life cycle methods

**Comment by xydian** [comment]
@EhsanSarshar its not that easy unfortunately, read the entire thread for details

**Comment by malvinder** [comment]
Confirming, still an issue in RN 0.60.4

**Comment by abdrmdn** [comment]
Confirming, still an issue in RN 0.59.0

**Comment by harrisonschwartz** [comment]
We're running into this issue as well.

**Comment by acoutts** [comment]
I just upgraded to 0.61.2 and the new fast refresh feature doesn't have this issue. Curious if anyone else can confirm as well.

**Comment by danielkv** [comment]
No solution for this yet?

**Comment by jedashford** [comment]
This is also an issue for those of us using codepush. When the app refreshes the js bundle, the previous state modal still shows and completely locks the app.

**Comment by vjsingh** [comment]
Very frustrating, even more so if you don't know what's causing it.

**Comment by hongdung6992** [comment]
Confirming, still an issue in RN 0.61.5
23/12/2019

**Comment by sunnylqm** [comment]
@hongdung6992 The fix will land on 0.62-rc1

**Comment by finntenzor** [comment]
A compatibility way to fix this problem if you have to use RN 0.59~0.61.(At least work for me)

Extend the ReactModalHostView and override the onHostPause method. Then use the new class instead of origin one. You may have to copy many Java files and js files from source code.

```java
// Fix the ReactModalHostView
public class ReactModalHostFixedView extends ReactModalHostView {
    private Method dismissMethod;

    public ReactModalHostFixedView(Context context) {
        super(context);
    }

    @Override
    public void onHostPause() {
        super.onHostPause();
        Method dismiss = this.getDismissMethod();
        try {
            dismiss.invoke(this);
        } catch (InvocationTargetException e) {
            Log.e("Demo", "Cant invoke dismiss(InvocationTargetException )");
        } catch (IllegalAccessException e) {
            Log.e("Demo", "Cant invoke dismiss(IllegalAccessException )");
        }
    }

    protected Method getDismissMethod() {
        if (dismissMethod == null) {
            try {
                Method dismiss = ReactModalHostView.class.getDeclaredMethod("dismiss");
                dismiss.setAccessible(true);
                dismissMethod = dismiss;
            } catch (NoSuchMethodException e) {
                Log.e("Demo", "No dismiss method");
            }
        }
        return dismissMethod;
    }
}
```

```java
// Fix the ReactModalHostManager 
public class ReactModalHostFixedManager extends ReactModalHostManager {
    public static final String REACT_CLASS = "RCTModalHostFixedView";

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected ReactModalHostView createViewInstance(ThemedReactContext reactContext) {
        return new ReactModalHostFixedView(reactContext);
    }
}
```

```java
// Create you own package and register it in "MainApplication.java"
// See https://facebook.github.io/react-native/docs/native-components-android
public class ReactModalHostFixedPackage implements  ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext)
    {
        retrun Collections.emptyList();
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext)
    {
        List<ViewManager> managers = new ArrayList<>();
        managers.add(new ReactModalHostFixedManager());
        return managers;
    }
}
```

```js
// You should copy the RCTModalHostViewNativeComponent.js to your project
// from node_modules/react-native/Libraries/Modal/RCTModalHostViewNativeComponent.js

// The preceding code was omitted
// Notice you have to rewrite the import statements
export default (codegenNativeComponent<NativeProps>('ModalHostView', {
  interfaceOnly: true,
  paperComponentName: 'RCTModalHostFixedView',
}): NativeComponentType<NativeProps>);
```

```js
// Use fixed view if the platform is android
// The other codes was omitted

import RCTModalHostOriginView from 'react-native/Libraries/Modal/RCTModalHostViewNativeComponent.js';
import RCTModalHostFixedView from './RCTModalHostViewNativeComponent.js';
const RCTModalHostView = Platform.OS === 'ios' ? RCTModalHostOriginView : RCTModalHostView;
```

It works for me.(Only test on android simulator)

**Comment by finntenzor** [comment]
A compatibility way to fix this problem if you have to use RN 0.60~0.61.(At least work for me)

Extend the ReactModalHostView and override the onHostPause method. Then use the new class instead of origin one. You may have to copy many Java files and js files from source code.

```java
// Fix the ReactModalHostView
public class ReactModalHostFixedView extends ReactModalHostView {
    private Method dismissMethod;

    public ReactModalHostFixedView(Context context) {
        super(context);
    }

    @Override
    public void onHostPause() {
        super.onHostPause();
        Method dismiss = this.getDismissMethod();
        try {
            dismiss.invoke(this);
        } catch (InvocationTargetException e) {
            Log.e("Demo", "Cant invoke dismiss(InvocationTargetException )");
        } catch (IllegalAccessException e) {
            Log.e("Demo", "Cant invoke dismiss(IllegalAccessException )");
        }
    }

    protected Method getDismissMethod() {
        if (dismissMethod == null) {
            try {
                Method dismiss = ReactModalHostView.class.getDeclaredMethod("dismiss");
                dismiss.setAccessible(true);
                dismissMethod = dismiss;
            } catch (NoSuchMethodException e) {
                Log.e("Demo", "No dismiss method");
            }
        }
        return dismissMethod;
    }
}
```

```java
// Fix the ReactModalHostManager 
public class ReactModalHostFixedManager extends ReactModalHostManager {
    public static final String REACT_CLASS = "RCTModalHostFixedView";

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected ReactModalHostView createViewInstance(ThemedReactContext reactContext) {
        return new ReactModalHostFixedView(reactContext);
    }
}
```

```java
// Create you own package and register it in "MainApplication.java"
// See https://facebook.github.io/react-native/docs/native-components-android
public class ReactModalHostFixedPackage implements  ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext)
    {
        return Collections.emptyList();
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext)
    {
        List<ViewManager> managers = new ArrayList<>();
        managers.add(new ReactModalHostFixedManager());
        return managers;
    }
}
```

```js
// You should copy the RCTModalHostViewNativeComponent.js to your project
// from node_modules/react-native/Libraries/Modal/RCTModalHostViewNativeComponent.js

// The preceding codes were omitted
// Notice you have to rewrite the import statements

// Use fixed view if the platform is android
const Platform = require('react-native/Utilities/Platform');
const paperComponentName =
  Platform.OS === 'android' ? 'RCTModalHostFixedView' : 'RCTModalHostView';

// export the fixed view
// NOTICE: for RN 0.61.5
export default (codegenNativeComponent<NativeProps>('ModalHostView', {
  interfaceOnly: true,
  paperComponentName,
}): NativeComponentType<NativeProps>);
```


```js
// You should copy the RCTModalHostViewNativeComponent.js to your project
// from node_modules/react-native/Libraries/Modal/RCTModalHostViewNativeComponent.js

// The preceding codes were omitted
// Notice you have to rewrite the import statements

// Use fixed view if the platform is android
const Platform = require('react-native/Utilities/Platform');
const paperComponentName =
  Platform.OS === 'android' ? 'RCTModalHostFixedView' : 'RCTModalHostView';

// export the fixed view
// NOTICE: for RN 0.60.5
module.exports = ((requireNativeComponent(
  paperComponentName,
): any): RCTModalHostViewNativeType);
```

```js
// You should copy the Modal.js to your project
// from node_modules/react-native/Libraries/Modal/Modal.js

// Use fixed view above
// The other codes were omitted
import RCTModalHostView from './RCTModalHostViewNativeComponent.js';
```

It works for me.(Only test on android simulator and my android phone for RN 0.61.5 and RN 0.60.5)

**Comment by alzalabany** [comment]
same problem still in 61.5, i solved it somehow by adding key prop to modal.

```
<Modal key="anyname">
</Modal>
```

this fixed duplicate problem in android both emulator and devices.

**note**: same problem also exists in release version android, not just during debug mode.

**Comment by DibPuelma** [comment]
@alzalabany this does not solves the issue in my case :(

Anyone with another workaround? I am using Code Push and the app is stuck whenever a modal is shown and the code is reloaded

**Comment by DibPuelma** [comment]
I found a workaround. Just don't use the modal when the Platform.OS === 'android'. It loses some functionalities like the animation, but you can program does manually if needed. Here is some code I'm using

<TouchableHighlight
        style={{
          elevation: 5,
          zIndex: 10,
        }}
        disabled={!backdropAllowed}
        onPress={backdropAllowed ? onClose : () => {}}>
        <SafeAreaView style={{
          height: '100%',
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          backgroundColor: 'rgba(0, 0, 0, 0.6)',
        }}>
        {withSkip &&
          <View style={{
            width: '100%',
            alignItems: 'flex-end',
            position: 'relative',
            top: 10,
            right: 0,
          }}>
            <Button transparent
              onPress={handleSkip}>
              <Text style={{
                fontWeight: 'bold',
                color: colors.lightGray,
                fontSize: fontSizes.h4
              }}>
                OMITIR
              </Text>
            </Button>
          </View>
        }
          <TouchableWithoutFeedback
            disabled={!backdropAllowed} onPress={() => {}} >
            <KeyboardAvoidingView behavior="position" style={{width: '90%', marginBottom: 20}}>
              {children}
            </KeyboardAvoidingView>
          </TouchableWithoutFeedback>
        </SafeAreaView>
      </TouchableHighlight>

Hope it helps. If anyone needs more help or has any other ideas, just tag me.

**Comment by WouterFlorijn** [comment]
@sunnylqm do you know if it would be possible to merge this fix into 0.61 as well? And is there a time frame for a stable 0.62?

**Comment by sunnylqm** [comment]
@WouterFlorijn For release related questions, https://github.com/react-native-community/releases/issues

**Comment by TKY2048** [comment]
It's been almost another whole year and this hasn't been fixed!

**Comment by sunnylqm** [comment]
@TKY2048 It's fixed in 0.62

**Comment by Daniel-Griffiths** [comment]
For anyone else who is using Expo and is stuck on RN 0.61, this fixed this issues for me: 

```
<Modal ...yourProps>
  <TouchableWithoutFeedback onPressOut={(e) => {
    if (e.nativeEvent.locationY < 0) {
      handleModal(false)
    }}
  >
    <View>
      ...modal content
    </View>
  </TouchableWithoutFeedback>
</Modal>
```

The `TouchableWithoutFeedback` fixes the problem.
